### DISCUSSION TITLE
How to print data from structure zfs ??

### DISCUSSION THREAD

**Discussion by constantmanish** [bpftrace/bpftrace#2268] [Q&A]
Hey guys, can someone please help me include zfs headers in the bpftrace program and print an attribute inside struct.

**Comment by fbs** [comment]
Hi,

Including a header and accessing struct values is similar to how its done in C:

```
#include <myheader.h>

kprobe:??? { 
  $a = (struct mystruct *)arg0;
  printf("first field: %d\n", $a->somefield;
}
```

But I doubt this is what you really need.

Can you explain what you want to do and what part you're stuck on. Keep in mind that zfs is a complex bit of software so just printing a struct field might not make sense at all :)

**Reply by constantmanish** [reply]
HI @fbs , 
Including the header something link this 
```
#include <sys/zfs_file.h>
```
give me error on my ubuntu machine, I tried manually defining the struct in bpftrace code and use it similar described by you and it works, but  manually definining many struct like is quite troubling.
Should I need to download the source code manually and include those header files ??

**Reply by fbs** [reply]
Without seeing the error its hard to tell but zfs is complex so wouldnt be surprised if you need to do a lot more work.

Do you have btf data available for zfs?
